sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp

```diff
@@ -0,0 +1,389 @@
+                                    matrix_layout::dynamic> &src,
+                   multi_ptr<T, Space, IsDecorated> dst, size_t stride,
+                   Group &sg) {
+  const auto idx = sg.get_group_linear_id() * sg.get_local_range()[0] +
+                   sg.get_local_linear_id();
+
+  if constexpr (std::is_same_v<T, double>) {
+    const auto thread_x = idx % N;
+    const auto thread_y = idx / N;
+
+    if constexpr (Layout == matrix_layout::row_major) {
+      for (int i = 0; i < 4; ++i) {
+        const int d_idx = thread_x + i * 4 * stride + thread_y * stride;
+        dst[d_idx] = src.data[i];
+      }
+    } else {
+      for (int i = 0; i < 4; ++i) {
+        const int d_idx = i * 4 + thread_x * stride + thread_y;
+        dst[d_idx] = src.data[i];
+      }
+    }
+  } else if constexpr (std::is_same_v<T, float> || std::is_same_v<T, int32_t>) {
+    if constexpr (M == 16 && N == 16) {
+      const auto thread_x = idx % N;
+      const auto thread_y = idx / N;
+
+      if constexpr (Layout == matrix_layout::row_major) {
+        for (int i = 0; i < 4; ++i) {
+          const int d_idx = thread_x + i * stride + thread_y * 4 * stride;
+          dst[d_idx] = src.data[i];
+        }
+      } else {
+        for (int i = 0; i < 4; ++i) {
+          const int d_idx = i + thread_x * stride + thread_y * 4;
+          dst[d_idx] = src.data[i];
+        }
+      }
+    } else if constexpr (M == 32 && N == 32) {
+      const auto thread_x = idx % N;
+      const auto thread_y = idx / N;
+
+      if constexpr (Layout == matrix_layout::row_major) {
+        for (int j = 0; j < 4; ++j) {
+          for (int i = 0; i < 4; ++i) {
+            const int d_idx =
+                thread_x + i * stride + thread_y * 4 * stride + j * 8 * stride;
+            dst[d_idx] = src.data[i + 4 * j];
+          }
+        }
+      } else {
+        for (int j = 0; j < 4; ++j) {
+          for (int i = 0; i < 4; ++i) {
+            const int d_idx = i + thread_x * stride + thread_y * 4 + j * 8;
+            dst[d_idx] = src.data[i + 4 * j];
+          }
+        }
+      }
+    }
+  }
+}
+
+template <typename Group, typename T, size_t M, size_t N,
+          access::address_space Space, access::decorated IsDecorated>
+void joint_matrix_store_hip(joint_matrix_hip<T, matrix_use::accumulator, M, N,
+                                             matrix_layout::dynamic> &src,
+                            multi_ptr<T, Space, IsDecorated> dst, size_t stride,
+                            matrix_layout layout, Group &sg) {
+  if (matrix_layout::row_major == layout) {
+    store_layoutT<Group, matrix_layout::row_major>(src, dst, stride, sg);
+  } else {
+    store_layoutT<Group, matrix_layout::col_major>(src, dst, stride, sg);
+  }
+}
+
+template <typename Tm, typename Tc, std::size_t M, std::size_t K, std::size_t N,
+          matrix_layout LayoutA, matrix_layout LayoutB,
+          std::enable_if_t<(LayoutA == matrix_layout::row_major ||
```

Comment:
I don't think this `enable_if_t` is necessary: `joint_matrix` with `matrix_use::a` `matrix_use::b` cannot be constructed if layout != row/col major. So other cases will never occur, so the check will never be used.